### DISCUSSION TITLE
Improving Loops

### DISCUSSION THREAD

**Discussion by jonathantneal** [withastro/roadmap#94] [Proposal]
Looping over a list of items in Astro is an exercise in understanding JavaScript.

**Example: Astro / JSX**
```astro
<ul>
  {cats.map(cat => (
    <li>
      <a target="_blank" href={`https://www.youtube.com/watch?v={cat.id}`}>
        {cat.name}
      </a>
    </li>
  )}
</ul>
```

This is not the case in other languages:

**Example: Svelte**
```svelte
<ul>
  {#each cats as cat}
    <li>
      <a target="_blank" href="https://www.youtube.com/watch?v={cat.id}">
        {cat.name}
      </a>
    </li>
  {/each}
</ul>
```

As an author who prefers the _simplicity_ of HTML-like markup in Astro files, I would like to do this using **directives**.

**Hypothetical Example: Astro**
```astro
<ul loop:each="cats" loop:as="cat">
  <li>
    <a target="_blank" href={`https://www.youtube.com/watch?v={cat.id}`}>
      {cat.name}
    </a>
  </li>
</ul>
```

**Comment by boehs** [comment]
Can we create directives without hard coding it into astro?

**Reply by natemoo-re** [reply]
I'd like to avoid supporting arbitrary, user-defined directives for a few reasons:

1. It could be a major performance hit to check if every usage of XML namespaces is actually a directive or just a name-spaced attribute.
2. You should be able to grok what features are supported when switching between Astro projects. I don't want to accidentally lead us into a world where not only do users have to learn Astro, they also have to learn a particular project's _flavor_ of Astro.
3. I'm really not sure what information a directive API would need to expose... it seems to vary widely depending on use case. `set:html` and `set:text` are used to modify the actual AST, `is:raw` would modify the actual tokenization of a file, `client:*` instructs Astro to generate/bundle components for client-side usage, and others like `class:list` just transform the directive value.

**Comment by natemoo-re** [comment]
I have been wanting to tackle this for quite some time! 

Curious why you suggested directives rather than Components? Components seem like the right primitive to me since they are extensible and can be written by users. It might require _less_ work in the compiler to fix conditional rendering than it would to hard-code whatever directives we land on.

**Reply by jonathantneal** [reply]
Thanks for looking at this, Nate. I’m excited someone else is interested!

I suggested directives because:
1. Component names seemed more problematic to analyze outside of runtime.  Not necessarily impossible. We would just need a system where the component name from the import or the variable could be traced out to its usage. Does that seem right?
2. As directives, loops can be communicated more easily alongside their context. This usefulness would be comparable to `set:html` working for elements, components, and fragments.
